### PR TITLE
Add beginner friendly documentation

### PR DESCRIPTION
**Pull request by faassen**
As discussed in issue https://github.com/mobxjs/mobx-react-docz/issues/99
the documention at present shows a great deal of options but is not very friendly for beginners (or as a quick refresher). This is a first stab at such documentation.

I'm not happy with the name "Basics" but the name "Introduction" is already in use. Perhaps we can rename that to make this the real introduction.

I'm also not happy that I couldn't actually test the code I wrote without copying them all into a separate file; I haven't found a file inclusion mechanism yet for code snippets. So the inline code is
all untested at present. :(  

Docz (v2?) offers a Playground component which comes close, but focuses on React and I'm not sure what to do about model code then.

### CODE REVIEW

File: content/basics/basics.mdx [hunk 1]
```diff
@@ -0,0 +1,327 @@
+---
+name: Basics
```
Comment (FredyC) [inline content/basics/basics.mdx:2]: What about "Getting started"? :)

File: content/basics/basics.mdx [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+route: /basics
+---
+
+import { Playground } from 'docz'
+
+# Basic Introduction
+
+[MobX](https://mobx.js.org/README.html) lets you create observable state. This
```
Comment (FredyC) [inline content/basics/basics.mdx:10]: Link to the site directly in case we would restructure it in the future.
```suggestion
[MobX](https://mobx.js.org) lets you create observable state. This
```

File: content/basics/basics.mdx [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+library, `mobx-react` integrates it with React. It provides a way for you to
+create React components that respond to MobX state changes.
```
Comment (FredyC) [inline content/basics/basics.mdx:12]: The library isn't really creating anything, just wrapping.
```suggestion
enhance React components that respond to MobX state changes.
```

File: content/basics/basics.mdx [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+
+> Frameworks such as [mobx-state-tree](https://mobx-state-tree.js.org) and
+> [mobx-keystone](https://mobx-keystone.js.org/) build on
+> top of MobX to offer a higher-level way to create observable models. These models
+> give you additional properties, such as automatic serialization to JSON. This
+> introduction uses plain MobX models. The principles of React
+> integration are the same for these advanced libraries are the same
```
Comment (FredyC) [inline content/basics/basics.mdx:19]: This ending of this sentence doesn't make much sense, try to rewrite it differently.

File: content/basics/basics.mdx [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+> These code examples use TypeScript. If you use plain JavaScript you
+> can skip the type declarations.
+
+```tsx
+import { observable, action, computed } from 'mobx'
+
+class Animal {
+  name: string
+  species: string
+  @observable energy: number
```
Comment (FredyC) [inline content/basics/basics.mdx:37]: Given that upcoming MobX 6 is dropping support for decorators, let's not spread it here. Check https://github.com/mobxjs/mobx/issues/2325 for API examples that are most likely set in the stone by now.

File: content/basics/basics.mdx [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+## A React component that observes
+
+We haven't used anything except plain MobX in this example. Now let's
+create a React component to show information for an animal:
+
+```tsx
+import { Component } from 'react'
+import { observer } from 'mobx-react'
+
+const AnimalView = observer(({ animal }: { animal: Animal }) => {
```
Comment (FredyC) [inline content/basics/basics.mdx:83]: It's a matter of preference, but I think this is more readable as there is less animals together :)
```suggestion
type TProps = {
  animal: Animal
}
const AnimalView = observer<TProps>(({ animal }) => {
```

File: content/basics/basics.mdx [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+makes `AnimalView` automatically re-render itself if anything that it observes
+changes. In this case the observable information is the computed property
+`isHungry`, which in turn it based on the observable property `energy`.
+`AnimalView` automatically re-renders itself only if `isHungry` changes. You
+don't need to do anything else to make that happen.
+
+When you wrap your React component in `observer`, it tracks any observable
+property or object you access during rendering it, and reacts to any changes to
+these. It won't react to any changes in observable properties you _don't_
+directly or indirectly use during rendering.
```
Comment (FredyC) [inline content/basics/basics.mdx:121]: What would "indirectly" mean? Perhaps add an example for that?

File: content/basics/basics.mdx [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+import { observer } from 'mobx-react'
+
+// Uses observer() but does not benefit from observability
+const NonObservingAnimalView = observer(
+  ({
+    name,
+    species,
+    isHungry,
+  }: {
+    name: string
```
Comment (FredyC) [inline content/basics/basics.mdx:140]: Same here as above, I would separate TProps to make it more readable.

File: content/basics/basics.mdx [hunk 9]
```diff
@@ -0,0 +1,327 @@
...
+
+Note that MobX can only observe properties, not simple values like a `string`
+or a `number`. So if you write your component to only receive simple values, it
+does not benefit from observability:
+
+```tsx
+import { Component } from 'react'
+import { observer } from 'mobx-react'
+
+// Uses observer() but does not benefit from observability
```
Comment (FredyC) [inline content/basics/basics.mdx:133]: Have you tried it? I believe that destructuring props like is still fine because `observer` sees access to those properties.

File: content/basics/basics.mdx [hunk 10]
```diff
@@ -0,0 +1,327 @@
...
+)
+```
+
+The component `NonObservingAnimalView` works perfectly if it is used within
+another component that receives an `Animal` instance and is also an observer,
+but it re-render each and every time, because MobX cannot see whether
+`name`, `species` or `isHungry` are changed. Our original `AnimalView` _does_
+observe it, because it can track property accesses.
+
+Wrapping a component in `observer` is cheap, so if you use MobX with react,
```
Comment (FredyC) [inline content/basics/basics.mdx:160]:
```suggestion
Wrapping a component in `observer` is cheap, so if you use MobX with React,
```

File: content/basics/basics.mdx [hunk 11]
```diff
@@ -0,0 +1,327 @@
...
+it's recommended to just wrap all your components in `observer`, even if the
```
Comment (FredyC) [inline content/basics/basics.mdx:161]: I wouldn't say it's "recommended". It's an escape hatch so you don't need to think about it. The recommended is to use https://mobx.js.org/refguide/api.html#computedrequiresreaction-boolean

File: content/basics/basics.mdx [hunk 12]
```diff
@@ -0,0 +1,327 @@
...
+component does not use any observable information yet. There is no need
+distinguish between React components that observe and do not observe in this.
+
+## Observable arrays
+
+MobX does support observable arrays, observable
+objects and observable maps, so those _can_ be observed for changes by
+a React component. We demonstrate that here with `animals`:
```
Comment (FredyC) [inline content/basics/basics.mdx:169]:
```suggestion
MobX does support arrays, objects, Maps, and Sets. Those _can_ be observed for changes by a React component. We demonstrate that here with `animals`:
```

File: content/basics/basics.mdx [hunk 13]
```diff
@@ -0,0 +1,327 @@
...
+zoo.addAnimal(wilma)
+```
+
+We can now render our animals:
+
+```tsx
+<AnimalsView animals={zoo.animals} />
+```
+
+`AnimalsView` automatically re-renders itself whenever you change the array, in
```
Comment (FredyC) [inline content/basics/basics.mdx:205]: It would be worth to mention common pitfall here where people add `observer` around parent component rendering `AnimalsView` but to that component alone. In such a case, adding/removing from such array won't be reactive.

File: content/basics/basics.mdx [hunk 14]
```diff
@@ -0,0 +1,327 @@
...
+itself as before when any observed properties the underlying object change,
+which in this case is its `isHungry` status.
+
+## Local component state with `useLocalStore`
+
+So far we have discussed observable model state, where the models live outside
+of the React components. It is useful to have component-specific state however.
+This is state that only makes sense within that component, is typically very UI
+specific, and there is no reason to specify this part of your application's
+models.
```
Comment (FredyC) [inline content/basics/basics.mdx:216]:
```suggestion
This is a state that only makes sense within that component. It is typically very UI
specific, and there is no reason to specify this part of your application's
models.
```

File: content/basics/basics.mdx [hunk 15]
```diff
@@ -0,0 +1,327 @@
...
+
+This is useful for "temporary data", such as for instance the text that a user
```
Comment (FredyC) [inline content/basics/basics.mdx:218]: I don't like the way you are treating `useLocalStore` here. It's far from "temporary data" only. It's a regular observable you can pass around, including actions & computeds. You can even stick into the Context and use anywhere in the app. It's equivalent to class-based stores. I know, the name of that hook tells something else, but it's too late to change that now.

File: content/basics/basics.mdx [hunk 16]
```diff
@@ -0,0 +1,327 @@
...
+Here we define an object called `local` with two observable properties, `name`
+and `species`. We use the React [controlled
+components](https://reactjs.org/docs/forms.html#controlled-components)
+technique to allow the user to edit these properties. As you can see, this
+simply involves mutating the local state directly. We also have an `Add animal`
+button which uses the local state to modify the zoo using the `addAnimal`
+action. It then needs to reset the local state again to allow the user to
+enter another animal.
+
+XXX interaction with strict mode, see https://github.com/mobxjs/mobx-react-lite/issues/235
```
Comment (FredyC) [inline content/basics/basics.mdx:284]: XXX? :)

File: content/basics/basics.mdx [hunk 17]
```diff
@@ -0,0 +1,327 @@
...
+
+Where does `AnimalsView` get its `animals` prop from? From the `zoo` in
+`ZooView`. Where does `AnimalView` get its `animal` prop? From the each animal
+in the `AnimalsView`. We also pass `zoo` into `AddAnimalView` directly.
+
+But where would we put the `zoo`? One simple approach is to make it a global in
+your application, and simply import it when you need it.
+
+Often this is sufficient. Consider a `Locale` MobX model that describes information
+about your site's Locale. You can simply maintain the current locale as
```
Comment (FredyC) [inline content/basics/basics.mdx:318]: Let's warn here that global variables might be problematic for integration testing, it would cause flakiness of such tests.

File: content/basics/basics.mdx (hunk 3 above)
Comment (faassen) [reply content/basics/basics.mdx:12]: I've rewritten it to "a way for you to make React components respond to MobX state changes now" now.

File: content/basics/basics.mdx (hunk 15 above)
Comment (faassen) [reply content/basics/basics.mdx:218]: This is a tricky issue. You're right of course in that you can do anything with useLocalState. But I want to make sure the documentation doesn't offer too many options at once, and highlights the model-driven nature of many MobX applications. And I do want to mention the use case of React component specific state. I've added a bit of language to make clear it can do more for you too.

File: content/basics/basics.mdx (hunk 16 above)
Comment (faassen) [reply content/basics/basics.mdx:284]: Yes, a "TODO" comment. The problem is that this whole example breaks down, if I understand it correctly, as soon as you use MobX strict mode. On this issue you suggest useLocalStore is not worth it in strict mode. But what then would be the equivalent for the local component state use case? 

I want to write a document with *one* way to do it, and I either have to go to add text that says you shouldn't use MobX strict mode, or text that says, use MobX strict mode and this is what you use (but what? not useLocalStore).

File: content/basics/basics.mdx (hunk 17 above)
Comment (faassen) [reply content/basics/basics.mdx:318]: I think this section went against my own philosophy of "one way to do it" for this document by offering a lot of competing options, so I cut out the discussion of Context for now. I eventually want to add a section called "Context and cross-cutting concerns" but that can wait.

File: content/basics/basics.mdx (hunk 5 above)
Comment (faassen) [reply content/basics/basics.mdx:36]: Okay, I didn't know that and I need to look into that, thanks!

File: content/basics/basics.mdx (hunk 16 above)
Comment (faassen) [reply content/basics/basics.mdx:294]: In addition I just read that MobX 6 wants to make 'observed' be the default. So if that doesn't go together with `useLocalStore` we need to go back to the drawing board completely for this section.

**Comment by faassen** [comment]
The two most important todo items at this point are:

* move away from decorators

* figure out how `useLocalStore` works with MobX strict mode (or an alternative to `useLocalStore`)

I'd also like to add a section "Context and cross-cutting concerns" where I want to go into the use of React context a little.

File: content/getting_started/getting_started.mdx [hunk 18]
```diff
@@ -0,0 +1,332 @@
...
+      {animal.isHungry ? 'hungry' : 'not hungry'}.
+    </div>
+  )
+})
+```
+
+> These examples work with `mobx-react-lite` too. `mobx-react` expands
+> `mobx-react-lite` to have support for React class components. If you are
+> creating a new project that only uses React function components, you can use
+> `mobx-react-lite`.
```
Comment (FredyC) [inline content/getting_started/getting_started.mdx:95]:
```suggestion
> These examples are using `mobx-react` which builds on top of `mobx-react-lite`
> and adds support for React class components. If you are creating a new project
> that only uses React function components, you can use `mobx-react-lite`.
```

File: content/getting_started/getting_started.mdx [hunk 19]
```diff
@@ -0,0 +1,332 @@
...
+
+```tsx
+<AnimalView animal={fred} />
+```
+
+The only special thing that `AnimalView` does compared to a normal React
+function component is that it is wrapped in `observer` from `mobx-react`. This
+makes `AnimalView` automatically re-render itself if anything that it observes
+changes. In this case the observable information is the computed property
+`isHungry`, which in turn it based on the observable property `energy`.
```
Comment (FredyC) [inline content/getting_started/getting_started.mdx:113]:
```suggestion
`isHungry`, which in turn is based on the observable property `energy`.
```

File: content/basics/basics.mdx (hunk 17 above)
Comment (FredyC) [reply content/basics/basics.mdx]: Yea, context is certainly more advanced, and in simple cases passing through props works justice fine. But I would like to have here a clear notion that stores in global variables might bite you.

File: content/basics/basics.mdx (hunk 16 above)
Comment (FredyC) [reply content/basics/basics.mdx]: Yes, you are right. I think we will need to rethink `useLocalStore` anyway with MobX 6, it won't be that useful anymore. I will open a new issue to track that.

Comment (FredyC) [reply content/basics/basics.mdx]: Actually, we are already tracking this problem in https://github.com/mobxjs/mobx-react-lite/issues/259. It would be probably best to remove `useLocalStore` from this document completely because it's uncertain how it will look like in the end.

File: content/basics/basics.mdx (hunk 15 above)
Comment (FredyC) [reply content/basics/basics.mdx]: I agree. As I said in a comment below, we shouldn't mention `useLocalStore` here at all. Feel free to extract this portion to a separate file that won't be linked anywhere and we will see how it ends up.

File: content/basics/basics.mdx (hunk 5 above)
Comment (FredyC) [reply content/basics/basics.mdx]: Continuing here with [your concerns about decorators](https://github.com/mobxjs/mobx-react-docz/pull/102#issuecomment-646591780). I am not sure what is your confusion. The first comment shows "API changes" section on how it will look like with MobX 6. I know it's not technically available yet, but there is plenty of examples with decorators in other docs. Here I would like to target it more toward the future. We can even include some warning this API is not available yet. We can have it here like "promotion" on what's will become "next standard"

Comment (faassen) [reply content/basics/basics.mdx]: I am not interested in documenting an API that isn't available yet.

File: content/basics/basics.mdx (hunk 15 above)
Comment (faassen) [reply content/basics/basics.mdx]: This means that a very commonly used pattern with MobX has no equivalent with functional components. I mean this one:

```js
class MyReactComponent extends Component {
  @observable isOpen = false
}
```
 
Are functional components with hooks read for prime time?

**Comment by faassen** [comment]
I'm stuck with this PR now:

* I want to document decorator style observables, as that's the current best practice with MobX, documented in the MobX tutorials. I realize that is going to change, but this change hasn't landed yet and I want to document something that works today. I can add a note that this is going to change, but even the MobX documentation itself doesn't have such a note yet. We can always update this as soon as MobX 6 is released.

* I want to document the two major patterns: observable model classes, and local observable state in React components. But it turns out that if you use documented MobX best practices (which enforces actions, and indeed the future default behavior) there is no proper way to have local component state. I can omit the whole section, but I think this is a major flaw of functional components -- a project that uses class-based React has no upgrade path. I'm willing to split this discussion into a separate (hidden) document for now, though, but I'm rather taken aback by this.

**Comment by FredyC** [comment]
I see what you mean, we will discuss the upgrade path for `useLocalStore` for sure. I don't have a clear answer on that now. The `useLocalStore` was always a kinda long shot and we haven't propagated much as a recommended pattern because there are potential issues with React Concurrent mode.

Personally, after some experience, I am fairly convinced that MobX based local state is not that useful. These days I would rather pick Immer which is similar to MobX with mutability. There is effectively very little gain in having MobX as a local store unless passed down through props.

About decorators, as I said above, I would like to avoid those because it's documented in many other places. There is no point in making a new document that encourages deprecated patterns.

If you want to proceed, you can just as well [use `decorate`](https://mobx.js.org/refguide/api.html#decorate) instead which is much closer to the upcoming API.

**Comment by faassen** [comment]
I don't want to document the 'decorate' API. This is an API that will be deprecated as soon as the new constructor-based API is released in MobX 6, while like it or not, decorators have been promoted with MobX for years and are everywhere in the documentation and examples. If I suddenly say, okay, but now that you want to use MobX with React you suddenly have to switch to something else, I only cause confusion.

Concerning local component state: I think there's a benefit to having a way to work with state that works the same no matter where it comes from, though it might be minor; perhaps, for instance, a minor performance benefit if you use computed local component state that is combined with model-based state. Going to 'immer' instead sounds strange to me as suddenly we introduce a whole new state management solution, but I'm willing to consider recommending React's "useState"? I'd really like to hear more from the wider MobX developer community on this.

**Comment by FredyC** [comment]
> I don't want to document the 'decorate' API. This is an API that will be deprecated as soon as the new constructor-based API

You are contradicting yourself. The `decorate` will be deprecated in a same way as decorators. You cannot really say it's better to document decorators now for that reason. The major reason why you see decorators in all examples is its simplicity. It's always easier to use decorators and explain concepts on collocated code. Let's postpone the docs update until V6. As I said, I don't want to add decorators here and then remove it later. 

> Going to 'immer' instead sounds strange to me as suddenly we introduce a whole new state management solution

I guess you don't know what is Immer about? It's as far from any "state management" as it can get. It's only a _magical wand_ to make state updates easier - uses mutations instead of immutable structures, but inside it just `useState/useReducer`.

At this point, it's totally fine to recommend `useLocalStore` as it will be working in V6.

**Comment by faassen** [comment]
I'm not interested in this anymore.

**Comment by geohuz** [comment]
Is there anybody like me feel stupid when looking at the documentation? I found the examples are hard to follow as they can't run if you just copy/paste code in your project as they are not complete, some of the code are skipped in the examples,  I finally make some work after digging into the git repository example folders. Also, it looks like the author assumes that  typescript should be a standard and everyone should follow? It just add recongize load for people not familiar with it. 

The example codes are like excerpt, for example, the context usage, there is a StoreProvider declared but I don't see the use of StoreProvider.

I really appreciate the hard work of people dedicate into the opensource project, but if you really want people make a good use of it, please kindly provide a more friendly document.

Thanks

**Comment by FredyC** [comment]
@geohuz If you want to make things better, your help is welcome as well. I did my best to make those docs at the time. I definitely want to keep TypeScript there, but if you can come up with the way to have JS there as well, I am open to it.

**Comment by faassen** [comment]
@geohuz I recommend you look at the Mobx documentation itself, as well as the readme documentation; the documentation on the mobx-react website doesn't really work for me either.

https://mobx.js.org/refguide/observer-component.html

and possibly the README

https://github.com/mobxjs/mobx-react
